zerver/worker/queue_processors.py

```diff
@@ -607,18 +612,35 @@ def maybe_send_batched_emails(self) -> None:
 
 
 @assign_queue("email_senders")
-class EmailSendingWorker(QueueProcessingWorker):
+class EmailSendingWorker(LoopQueueProcessingWorker):
+    def __init__(self) -> None:
+        super().__init__()
+        self.connection: EmailBackend = initialize_connection(None)
+
     @retry_send_email_failures
-    def consume(self, event: Dict[str, Any]) -> None:
+    def send_email(self, event: Dict[str, Any]) -> None:
         # Copy the event, so that we don't pass the `failed_tries'
-        # data to send_email_from_dict (which neither takes that
-        # argument nor needs that data).
+        # data to send_email (which neither takes that argument nor
+        # needs that data).
         copied_event = copy.deepcopy(event)
         if "failed_tries" in copied_event:
             del copied_event["failed_tries"]
-        handle_send_email_format_changes(copied_event)
         send_email_from_dict(copied_event)
 
+    def consume_batch(self, events: List[Dict[str, Any]]) -> None:
+        for event in events:
+            # Refresh connection if needed
+            self.connection = initialize_connection(self.connection)
+            event["connection"] = self.connection
```

Comment:
This line is necessary in my opinion as the connection attribute is needed by the `send_email` function in `send_email.py` to actually send the email through it. However it is not at the right place. The issue that you mentioned above is linked to the `retry_send_email_failures` decorator that pushes the event back to the queue three times before giving up. This means that we can't insert the connection attribute in the event but rather have to copy it before passing it to `send_email` with the connection attribute. As the event was already copied to avoid passing `failed_tries` to `send_email_from_dict` I used the same copy to include the connection without impacting the serialization of the event that will be pushed into the queue.